modules/core/src/smithy4s/Endpoint.scala

```diff
@@ -63,75 +63,114 @@ object Endpoint {
     def errorable: Option[Errorable[E]]
     def streamedInput: StreamingSchema[SI]
     def streamedOutput: StreamingSchema[SO]
+    def builder: Builder[I, E, O, SI, SO] = Builder.fromEndpoint(this)
+  }
+
+  trait Middleware[A] { self =>
+    def prepare[Alg[_[_, _, _, _, _]]](service: Service[Alg])(endpoint: service.Endpoint[_, _, _, _, _]): A => A
+    final def biject[B](to: A => B)(from: B => A): Middleware[B] = new Middleware[B] {
+      def prepare[Alg[_[_, _, _, _, _]]](service: Service[Alg])(endpoint: service.Endpoint[_, _, _, _, _]): B => B =
+        self.prepare(service)(endpoint).compose(from).andThen(to)
+    }
+
+    final def andThen(other: Middleware[A]): Middleware[A] =
+      new Middleware[A] {
+        def prepare[Alg[_[_, _, _, _, _]]](
+            service: Service[Alg]
+        )(endpoint: service.Endpoint[_, _, _, _, _]): A => A =
+          self
+            .prepare(service)(endpoint)
+            .andThen(other.prepare(service)(endpoint))
+      }
+
+  }
+// format: on
+
+  object Middleware {
+
+    trait Simple[Construct] extends Middleware[Construct] {
+      def prepareWithHints(serviceHints: Hints, endpointHints: Hints): Construct => Construct
+
+      final def prepare[Alg[_[_, _, _, _, _]]](service: Service[Alg])(
+          endpoint: service.Endpoint[_, _, _, _, _]
+      ): Construct => Construct =
+        prepareWithHints(service.hints, endpoint.hints)
+    }
+
+    def noop[Construct]: Middleware[Construct] =
+      new Middleware[Construct] {
+        override def prepare[Alg[_[_, _, _, _, _]]](service: Service[Alg])(
+            endpoint: service.Endpoint[_, _, _, _, _]
+        ): Construct => Construct = identity
+      }
 
   }
 
   type ForOperation[Op[_, _, _, _, _]] = {
     type e[I, E, O, SI, SO] = Endpoint[Op, I, E, O, SI, SO]
   }
 
-  final case class Builder[Op[_, _, _, _, _], I, E, O, SI, SO] private (
+  final case class Builder[I, E, O, SI, SO] private (
```

Comment:
@kubukoz, so one thing that's appearing more and more obvious is that the coupling of `Endpoint.Base` is a very valuable construct on its own, and it's annoying that `Endpoint` uses inheritance instead of composition. So I think it's gonna be important to decouple `Endpoint` with `Endpoint.Base`, which also implies decoupling the builders. 

I think this is one of the things that will be extracted in a separate PR